src/arch/arm/64/kernel/vspace.c

```diff
@@ -1080,51 +1080,36 @@ pude_t *pageDirectoryMapped(asid_t asid, vptr_t vaddr, pde_t *pd)
 
 #ifdef CONFIG_ARM_HYPERVISOR_SUPPORT
 
-static void invalidateASID(asid_t asid)
+static asid_map_t *findMapRefForASID(asid_t asid)
 {
-    asid_pool_t *asidPool;
-
-    asidPool = armKSASIDTable[asid >> asidLowBits];
-    assert(asidPool);
-
-    asid_map_t asid_map = asidPool->array[asid & MASK(asidLowBits)];
-    assert(asid_map_get_type(asid_map) == asid_map_asid_map_vspace);
+    asid_pool_t        *poolPtr;
 
-    vspace_root_t *vtable = (vspace_root_t *)asid_map_asid_map_vspace_get_vspace_root(asid_map);
+    poolPtr = armKSASIDTable[asid >> asidLowBits];
+    assert(poolPtr != NULL);
 
-    vtable[VTABLE_VMID_SLOT] = vtable_invalid_new(0, false);
+    return &poolPtr->array[asid & MASK(asidLowBits)];
 }
 
-static vspace_root_t PURE loadHWASID(asid_t asid)
+static void invalidateASID(asid_t asid)
 {
-    asid_pool_t *asidPool;
-
-    asidPool = armKSASIDTable[asid >> asidLowBits];
-    assert(asidPool);
+    asid_map_t *asid_map;
 
-    asid_map_t asid_map = asidPool->array[asid & MASK(asidLowBits)];
-    assert(asid_map_get_type(asid_map) == asid_map_asid_map_vspace);
+    asid_map = findMapRefForASID(asid);
+    assert(asid_map_get_type(*asid_map) == asid_map_asid_map_vspace);
 
-    vspace_root_t *vtable = (vspace_root_t *)asid_map_asid_map_vspace_get_vspace_root(asid_map);
-
-    return vtable[VTABLE_VMID_SLOT];
+    asid_map_asid_map_vspace_ptr_set_hw_asid(asid_map, 0);
+    asid_map_asid_map_vspace_ptr_set_hw_asid_valid(asid_map, false);
 }
 
 static void storeHWASID(asid_t asid, hw_asid_t hw_asid)
 {
-    asid_pool_t *asidPool;
-
-    asidPool = armKSASIDTable[asid >> asidLowBits];
-    assert(asidPool);
+    asid_map_t *asid_map;
 
-    asid_map_t asid_map = asidPool->array[asid & MASK(asidLowBits)];
-    assert(asid_map_get_type(asid_map) == asid_map_asid_map_vspace);
+    asid_map = findMapRefForASID(asid);
+    assert(asid_map_get_type(*asid_map) == asid_map_asid_map_vspace);
 
-    vspace_root_t *vtable = (vspace_root_t *)asid_map_asid_map_vspace_get_vspace_root(asid_map);
-
-    /* Store HW VMID in the last entry
-       Masquerade as an invalid PDGE */
-    vtable[VTABLE_VMID_SLOT] = vtable_invalid_new(hw_asid, true);
```

Comment:
seL4_UserTop can be updated as VTABLE_VMID_SLOT is no longer a removed slot.  Additionally, 1 slot at the end of the vspace is used to hold the number of SMMU CBs assigned to the vspace and this should also be added to the asid_map.